File: tempestweb/native/tabs.py
"""Native cross-tab capability over the BroadcastChannel and Web Locks Web APIs.

:func:`broadcast` sends a ``tabs.broadcast`` ``native_call`` and :func:`lock` /
:func:`unlock` manage named locks; ``client/native/tabs.js`` posts on a
``BroadcastChannel`` and drives ``navigator.locks``. Receiving broadcasts is a
continuous stream and is deferred to the event channel rather than a
request/response call. A lock held across the ``lock`` / ``unlock`` pair is kept
alive by the client on a registry keyed by its name.
"""

from __future__ import annotations

from tempestweb.native.dispatch import send_native_call

__all__ = ["broadcast", "lock", "unlock"]


async def broadcast(channel: str, message: object) -> None:
    """Post a message to all tabs listening on a broadcast channel.

    Args:
        channel: The broadcast channel name.
        message: The JSON-able message to post to other tabs.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    await send_native_call(
        "tabs.broadcast",
        {"channel": channel, "message": message},
    )


async def lock(name: str, mode: str = "exclusive") -> bool:
    """Acquire a named cross-tab lock.

    Args:
        name: The lock name.
        mode: The lock mode, ``"exclusive"`` or ``"shared"``.

    Returns:
        ``True`` if the lock was acquired.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("tabs.lock", {"name": name, "mode": mode})
    return bool(value.get("acquired", False))


async def unlock(name: str) -> None:
    """Release a previously acquired named cross-tab lock.

    Args:
        name: The lock name passed to :func:`lock`.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    await send_native_call("tabs.unlock", {"name": name})

File: client/native/tabs.js
// native/tabs.js — cross-tab coordination: BroadcastChannel + Web Locks.
//
// `broadcast` is a one-shot post over a named BroadcastChannel. `lock`/`unlock`
// model a held Web Lock: the Web Locks API only holds a lock for the lifetime of
// a callback promise, so we grant the lock, resolve the caller with
// `{acquired:true}` the moment it is held, and keep the callback's promise pending
// until `unlock` fires the stored release resolver. We hold those resolvers in a
// module-level registry keyed by lock name.

import { CapabilityError } from "./index.js";

/**
 * Release resolvers for held locks, keyed by lock name.
 * @type {Map<string, () => void>}
 */
const _releasers = new Map();

/**
 * Post a message to all tabs listening on a named BroadcastChannel.
 * @param {{channel:string, message:*}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object>}
 * @throws {CapabilityError} unavailable when BroadcastChannel is missing.
 */
export async function tabsBroadcast(args, deps) {
  const BroadcastChannelCtor =
    deps.BroadcastChannel || /** @type {any} */ (globalThis).BroadcastChannel;
  if (!BroadcastChannelCtor) {
    throw new CapabilityError("unavailable", "BroadcastChannel is not available");
  }
  const bc = new BroadcastChannelCtor(args.channel);
  bc.postMessage(args.message);
  bc.close();
  return {};
}

/**
 * Acquire a named Web Lock, holding it until {@link tabsUnlock} releases it.
 * @param {{name:string, mode?:string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{acquired:boolean}>}
 * @throws {CapabilityError} unavailable when the Web Locks API is missing.
 */
export async function tabsLock(args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.locks || typeof nav.locks.request !== "function") {
    throw new CapabilityError("unavailable", "the Web Locks API is not available");
  }
  const mode = args.mode || "exclusive";
  return await new Promise((resolveAcquired) => {
    nav.locks.request(
      args.name,
      { mode },
      () =>
        new Promise((release) => {
          _releasers.set(args.name, release);
          resolveAcquired({ acquired: true });
        }),
    );
  });
}

/**
 * Release a previously acquired named Web Lock (idempotent).
 * @param {{name:string}} args
 * @param {import("./index.js").NativeDeps} _deps
 * @returns {Promise<Object>}
 */
export async function tabsUnlock(args, _deps) {
  const release = _releasers.get(args.name);
  if (release) {
    release();
    _releasers.delete(args.name);
  }
  return {};
}
